vkubelet/pod.go

```diff
@@ -28,60 +30,87 @@ func addPodAttributes(ctx context.Context, span trace.Span, pod *corev1.Pod) con
 }
 
 func (s *Server) createOrUpdatePod(ctx context.Context, pod *corev1.Pod, recorder record.EventRecorder) error {
-	// Check if the pod is already known by the provider.
-	// NOTE: Some providers return a non-nil error in their GetPod implementation when the pod is not found while some other don't.
-	// Hence, we ignore the error and just act upon the pod if it is non-nil (meaning that the provider still knows about the pod).
-	if pp, _ := s.provider.GetPod(ctx, pod.Namespace, pod.Name); pp != nil {
-		// The pod has already been created in the provider.
-		// Hence, we return since pod updates are not yet supported.
-		log.G(ctx).Warnf("skipping update of pod %s as pod updates are not supported", pp.Name)
-		return nil
-	}
 
 	ctx, span := trace.StartSpan(ctx, "createOrUpdatePod")
 	defer span.End()
 	addPodAttributes(ctx, span, pod)
 
-	if err := populateEnvironmentVariables(ctx, pod, s.resourceManager, recorder); err != nil {
-		span.SetStatus(ocstatus.FromError(err))
-		return err
-	}
-
 	ctx = span.WithFields(ctx, log.Fields{
 		"pod":       pod.GetName(),
 		"namespace": pod.GetNamespace(),
 	})
 
-	if origErr := s.provider.CreatePod(ctx, pod); origErr != nil {
-		podPhase := corev1.PodPending
-		if pod.Spec.RestartPolicy == corev1.RestartPolicyNever {
-			podPhase = corev1.PodFailed
-		}
+	if err := populateEnvironmentVariables(ctx, pod, s.resourceManager, recorder); err != nil {
+		span.SetStatus(ocstatus.FromError(err))
+		return err
+	}
 
-		pod.ResourceVersion = "" // Blank out resource version to prevent object has been modified error
-		pod.Status.Phase = podPhase
-		pod.Status.Reason = podStatusReasonProviderFailed
-		pod.Status.Message = origErr.Error()
-
-		logger := log.G(ctx).WithFields(log.Fields{
-			"podPhase": podPhase,
-			"reason":   pod.Status.Reason,
-		})
-
-		_, err := s.k8sClient.CoreV1().Pods(pod.Namespace).UpdateStatus(pod)
-		if err != nil {
-			logger.WithError(err).Warn("Failed to update pod status")
-		} else {
-			logger.Info("Updated k8s pod status")
+	// Check if the pod is already known by the provider.
+	// NOTE: Some providers return a non-nil error in their GetPod implementation when the pod is not found while some other don't.
+	// Hence, we ignore the error and just act upon the pod if it is non-nil (meaning that the provider still knows about the pod).
+	if pp, _ := s.provider.GetPod(ctx, pod.Namespace, pod.Name); pp != nil {
+		// Pod Update Only Permits update of:
+		// - `spec.containers[*].image`
+		// - `spec.initContainers[*].image`
+		// - `spec.activeDeadlineSeconds`
+		// - `spec.tolerations` (only additions to existing tolerations)
+		// compare the hashes of the pod specs to see if the specs actually changed
+		expected := hashPodSpec(pp.Spec)
```

Comment:
Couldn't this get into a weird situation where if the provider does something like async updates that the update function is called, but not applied immediately? Shouldn't we try to dedup the updates (i.e. keep track of the last update we did?)

Reply:
I'm not sure that state keeping makes sense in the Kubernetes world. The API server is the source of truth / keeper of state. If we get an update and the spec has changed, that should signify that another update needs to be made by the provider. This is essentially the same behavior that occurs in a regular kubelet, except at a little different granularity (i.e. the kubelet diffs each container within the update). In the case of updating a pod, it's actually quite limited in what you can change: image (the containers and the init containers), tolerations, and active deadline. I think the more common thing to occur would be something like with a replica set or a deployment, which would result in pod create/deletes and not updates. Adding de-dupe on these seems like adding complexity to VK for a possibly infrequently used case. Providers might implement that logic if it makes sense. 

@cpuguy83 any thoughts?

Reply:
Agree that we don't want to keep state like this.
The reason being is that the k8s API is a declarative API, if a pod state updates 5x before we ever apply any changes, we should only apply the last state as it is the current expected state.
This is why the queue is setup to prefer storing keys to storing point-in-time objects.